src/jit/objectalloc.cpp

```diff
@@ -16,57 +16,227 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 {
+    JITDUMP("\n*** ObjectAllocationPhase: ");
     if ((comp->optMethodFlags & OMF_HAS_NEWOBJ) == 0)
     {
+        JITDUMP("no newobjs in this method; punting\n");
         return;
     }
 
     if (IsObjectStackAllocationEnabled())
     {
+        JITDUMP("enabled, analyzing...\n");
         DoAnalysis();
     }
+    else
+    {
+        JITDUMP("disabled, punting\n");
+    }
 
-    MorphAllocObjNodes();
+    const bool didStackAllocate = MorphAllocObjNodes();
+
+    if (didStackAllocate)
+    {
+        RewriteUses();
+    }
+}
+
+struct ObjectAllocator::BuildConnGraphVisitorCallbackData
+{
+    BuildConnGraphVisitorCallbackData(ObjectAllocator* objectAllocator) : m_objectAllocator(objectAllocator)
+    {
+    }
+
+    ObjectAllocator* m_objectAllocator;
+};
+
+//------------------------------------------------------------------------------
+// MarkLclVarAsEscaping : Mark local variable as escaping.
+//
+//
+// Arguments:
+//    lclNum  - Escaping pointing local variable number
+
+void ObjectAllocator::MarkLclVarAsEscaping(unsigned int lclNum)
+{
+    BitVecOps::AddElemD(&m_bitVecTraits, m_EscapingPointers, lclNum);
+}
+
+//------------------------------------------------------------------------------
+// IsLclVarEscaping : Check if the local variable has been marked as escaping.
+//
+//
+// Arguments:
+//    lclNum  - Local variable number
+//
+// Return Value:
+//    True if the local variable has been marked as escaping; false otherwise
+
+bool ObjectAllocator::IsLclVarEscaping(unsigned int lclNum)
+{
+    return BitVecOps::IsMember(&m_bitVecTraits, m_EscapingPointers, lclNum);
+}
+
+//------------------------------------------------------------------------------
+// AddConnGraphEdge : Record that the source local variable may point to the same set of objects
+//                    as the set pointed to by target local variable.
+//
+// Arguments:
+//    sourceLclNum  - Local variable number of the edge source
+//    targetLclNum  - Local variable number of the edge target
+
+void ObjectAllocator::AddConnGraphEdge(unsigned int sourceLclNum, unsigned int targetLclNum)
+{
+    BitVecOps::AddElemD(&m_bitVecTraits, m_ConnGraphAdjacencyMatrix[sourceLclNum], targetLclNum);
 }
 
 //------------------------------------------------------------------------
 // DoAnalysis: Walk over basic blocks of the method and detect all local
 //             variables that can be allocated on the stack.
-//
-// Assumptions:
-//    Must be run after the dominators have been computed (we need this
-//    information to detect loops).
+
 void ObjectAllocator::DoAnalysis()
 {
     assert(m_IsObjectStackAllocationEnabled);
-    assert(comp->fgDomsComputed);
-    // TODO-ObjectStackAllocation
-    NYI("DoAnalysis");
+    assert(!m_AnalysisDone);
+
+    if (comp->lvaCount > 0)
+    {
+        m_EscapingPointers         = BitVecOps::MakeEmpty(&m_bitVecTraits);
+        m_ConnGraphAdjacencyMatrix = new (comp->getAllocator()) BitSetShortLongRep[comp->lvaCount];
```

Comment:
It would be good to add a new CMK_ value for ObjectAllocator.

Reply:
Done.